pg_search/src/postgres/scan.rs

```diff
@@ -62,24 +63,98 @@ pub extern "C-unwind" fn amrescan(
-        match ScanStrategy::try_from(key.sk_strategy).expect("`key.sk_strategy` is unrecognized") {
-            ScanStrategy::TextQuery => unsafe {
-                let query_string = String::from_datum(key.sk_argument, false)
-                    .expect("ScanKey.sk_argument must not be null");
-                SearchQueryInput::Parse {
-                    query_string,
-                    lenient: None,
-                    conjunction_mode: None,
+        let strategy =
+            ScanStrategy::try_from(key.sk_strategy).expect("`key.sk_strategy` is unrecognized");
+
+        match strategy {
+            ScanStrategy::TextQuery => {
+                if let Some(query_string) = decode_text_value(key.sk_argument) {
+                    SearchQueryInput::Parse {
+                        query_string,
+                        lenient: None,
+                        conjunction_mode: None,
+                    }
+                } else if let Some(strings) = decode_text_array(key.sk_argument) {
+                    let should = strings
+                        .into_iter()
+                        .map(|query_string| SearchQueryInput::Parse {
+                            query_string,
+                            lenient: None,
+                            conjunction_mode: None,
+                        })
+                        .collect();
+                    disjunction_from_queries(should)
+                } else {
+                    panic!("failed to decode text-based ScanKey argument")
                 }
-            },
-            ScanStrategy::SearchQueryInput => unsafe {
-                SearchQueryInput::from_datum(key.sk_argument, false)
-                    .expect("ScanKey.sk_argument must not be null")
-            },
+            }
+            ScanStrategy::SearchQueryInput => {
+                if let Some(value) = decode_searchqueryinput_value(key.sk_argument) {
+                    value
+                } else if let Some(should) = decode_searchqueryinput_array(key.sk_argument) {
+                    disjunction_from_queries(should)
+                } else if let Some(strings) = decode_text_array(key.sk_argument) {
+                    let should = strings
+                        .into_iter()
+                        .map(|query_string| SearchQueryInput::Parse {
+                            query_string,
+                            lenient: None,
+                            conjunction_mode: None,
+                        })
+                        .collect();
+                    disjunction_from_queries(should)
+                } else {
+                    panic!("failed to decode SearchQueryInput ScanKey argument")
+                }
+            }
         }
     }
 
+    fn decode_searchqueryinput_value(datum: pg_sys::Datum) -> Option<SearchQueryInput> {
+        unsafe { SearchQueryInput::from_datum_resilient(datum, false) }
+    }
```

Comment:
This should not be made safe, because it is a direct call to an `unsafe` function without any checks to bar it from misuse.

Excusing such on the basis of module privacy is something I have seen done, but usually bites you later.